File: backends/kotlin-ktor/src/main/kotlin/com/authorization/usecase/gate/Interactor.kt
/*
 * ゲートユースケース Interactor モジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
package com.authorization.usecase.gate

import com.authorization.config.Config
import com.authorization.domain.client.JwtHistoryRepository
import com.authorization.domain.client.Repository as ClientRepository
import com.authorization.domain.gate.CacheRepository
import com.authorization.domain.gate.IssueVo
import com.authorization.domain.gate.VerifyVo
import com.nimbusds.jose.JWSAlgorithm
import com.nimbusds.jose.JWSHeader
import com.nimbusds.jose.crypto.RSASSASigner
import com.nimbusds.jose.crypto.RSASSAVerifier
import com.nimbusds.jwt.JWTClaimsSet
import com.nimbusds.jwt.SignedJWT
import java.security.KeyFactory
import java.security.interfaces.RSAPrivateKey
import java.security.interfaces.RSAPublicKey
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import java.time.Instant
import java.time.LocalDateTime
import java.util.Base64
import java.util.Date
import java.util.UUID

/**
 * ゲート認可ユースケースの Interactor です。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
class Interactor(
    private val clientRepo: ClientRepository,
    private val cache: CacheRepository,
    private val cfg: Config,
    private val historyRepo: JwtHistoryRepository? = null,
) {

    /**
     * クライアントのアクセストークンを検証して JWT を発行します。
     *
     * @param dto 発行 DTO
     * @return 発行結果 VO
     */
    suspend fun issueToken(dto: IssueDto): IssueVo {
        val c = clientRepo.findByAccessToken(dto.accessToken)
            ?: error("client_not_found")

        val cached = cache.getJwt(c.identifier, dto.memberId)
        if (!cached.isNullOrBlank()) return IssueVo(token = cached)

        val token = issueJwt(
            memberId      = dto.memberId,
            identifier    = c.identifier,
            privateKeyPem = c.privateKey,
            fingerprint   = c.fingerprint,
            issuer        = cfg.jwt.issuer,
            ttl           = cfg.jwt.ttl,
        )
        runCatching { cache.putJwt(c.identifier, dto.memberId, token, cfg.jwt.cacheTtl) }
        historyRepo?.let { runCatching { it.save(c.id, dto.memberId, LocalDateTime.now(), token) } }
        return IssueVo(token = token)
    }

    /**
     * JWT トークンを検証してクレームを返します。
     *
     * @param dto 検証 DTO
     * @return 検証結果 VO
     */
    suspend fun verify(dto: VerifyDto): VerifyVo {
        val c = clientRepo.findByIdentifier(dto.identifier)
            ?: error("client_not_found")
        val claims = verifyJwt(dto.identifier, dto.token, c.publicKey, cfg.jwt.issuer)
        return VerifyVo(claims = claims)
    }
}

private fun issueJwt(
    memberId: String,
    identifier: String,
    privateKeyPem: String,
    fingerprint: String,
    issuer: String,
    ttl: Long,
): String {
    val privateKey = loadPrivateKey(privateKeyPem)
    val now        = Instant.now()
    val claimsSet  = JWTClaimsSet.Builder()
        .issuer(issuer)
        .subject(memberId)
        .audience(identifier)
        .expirationTime(Date.from(now.plusSeconds(ttl)))
        .issueTime(Date.from(now))
        .notBeforeTime(Date.from(now))
        .jwtID(UUID.randomUUID().toString())
        .build()
    val header = JWSHeader.Builder(JWSAlgorithm.RS256).keyID(fingerprint).build()
    val jwt    = SignedJWT(header, claimsSet)
    jwt.sign(RSASSASigner(privateKey))
    return jwt.serialize()
}

private fun verifyJwt(
    identifier: String,
    tokenStr: String,
    publicKeyPem: String,
    issuer: String,
): Map<String, Any?> {
    val publicKey = loadPublicKey(publicKeyPem)
    val jwt       = SignedJWT.parse(tokenStr)
    check(jwt.verify(RSASSAVerifier(publicKey))) { "invalid_token" }
    val claims = jwt.jwtClaimsSet
    check(claims.issuer == issuer) { "invalid_issuer" }
    check(claims.audience.contains(identifier)) { "invalid_audience" }
    check(Date().before(claims.expirationTime)) { "token_expired" }
    return claims.toJSONObject().mapValues { it.value }
}

private fun loadPrivateKey(pem: String): RSAPrivateKey {
    val cleaned = pem
        .replace("-----BEGIN PRIVATE KEY-----", "")
        .replace("-----END PRIVATE KEY-----", "")
        .replace(Regex("\\s"), "")
    val bytes = Base64.getDecoder().decode(cleaned)
    return KeyFactory.getInstance("RSA").generatePrivate(PKCS8EncodedKeySpec(bytes)) as RSAPrivateKey
}

private fun loadPublicKey(pem: String): RSAPublicKey {
    val cleaned = pem
        .replace("-----BEGIN PUBLIC KEY-----", "")
        .replace("-----END PUBLIC KEY-----", "")
        .replace(Regex("\\s"), "")
    val bytes = Base64.getDecoder().decode(cleaned)
    return KeyFactory.getInstance("RSA").generatePublic(X509EncodedKeySpec(bytes)) as RSAPublicKey
}

File: backends/go-beego/internal/usecase/gate/interactor.go
// Package gate はゲートウェイ認証ユースケースを提供します。
package gate

import (
	domclient "authorization-go-beego/internal/domain/client"
	domgate "authorization-go-beego/internal/domain/gate"
	"authorization-go-beego/internal/config"
	"authorization-go-beego/pkg/apperror"
	"crypto/rsa"
	"crypto/x509"
	"encoding/pem"
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"
)

// Interactor はゲートウェイ認証ユースケースの実装です。
type Interactor struct {
	clientRepo  domclient.Repository
	historyRepo domclient.JwtHistoryRepository
	cache       domgate.CacheRepository
	cfg         *config.Config
}

// NewInteractor は Interactor を生成します。
func NewInteractor(
	clientRepo domclient.Repository,
	historyRepo domclient.JwtHistoryRepository,
	cache domgate.CacheRepository,
	cfg *config.Config,
) *Interactor {
	return &Interactor{clientRepo: clientRepo, historyRepo: historyRepo, cache: cache, cfg: cfg}
}

// IssueToken はアクセストークンを検証してJWTを発行します。
func (uc *Interactor) IssueToken(dto IssueDto) (*domgate.IssueVo, error) {
	c, err := uc.clientRepo.FindByAccessToken(&domclient.Client{AccessToken: dto.AccessToken})
	if err != nil || c == nil {
		return nil, apperror.Unauthorized("client_not_found")
	}

	identifier := c.Identifier
	cached, err := uc.cache.GetJwt(identifier, dto.MemberID)
	if err == nil && cached != "" {
		return &domgate.IssueVo{Token: cached}, nil
	}

	token, err := uc.issueJwt(dto.MemberID, identifier, c.PrivateKey, c.Fingerprint)
	if err != nil {
		return nil, err
	}

	_ = uc.cache.PutJwt(identifier, dto.MemberID, token, uc.cfg.JWT.CacheTTL)
	_ = uc.historyRepo.Save(c.ID, dto.MemberID, time.Now(), token)
	return &domgate.IssueVo{Token: token}, nil
}

// Verify はJWTトークンを検証してクレームを返します。
func (uc *Interactor) Verify(dto VerifyDto) (*domgate.VerifyVo, error) {
	c, err := uc.clientRepo.FindByIdentifier(&domclient.Client{Identifier: dto.Identifier})
	if err != nil || c == nil {
		return nil, apperror.Forbidden("client_not_found")
	}

	claims, err := uc.verifyJwt(dto.Identifier, dto.Token, c.PublicKey)
	if err != nil {
		return nil, err
	}
	return &domgate.VerifyVo{Claims: claims}, nil
}

// issueJwt はRSA秘密鍵でJWTを署名して返します。
func (uc *Interactor) issueJwt(memberID, identifier, privateKeyPEM, fingerprint string) (string, error) {
	privKey, err := parseRSAPrivateKey(privateKeyPEM)
	if err != nil {
		return "", fmt.Errorf("parse private key: %w", err)
	}

	now := time.Now()
	claims := jwt.MapClaims{
		"iss": uc.cfg.JWT.Issuer,
		"sub": memberID,
		"aud": []string{identifier},
		"exp": now.Add(time.Duration(uc.cfg.JWT.TTL) * time.Second).Unix(),
		"iat": now.Unix(),
		"nbf": now.Unix(),
		"jti": uuid.New().String(),
	}

	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
	token.Header["kid"] = fingerprint
	return token.SignedString(privKey)
}

// verifyJwt はRSA公開鍵でJWTを検証してクレームを返します。
func (uc *Interactor) verifyJwt(identifier, tokenStr, publicKeyPEM string) (map[string]interface{}, error) {
	pubKey, err := parseRSAPublicKey(publicKeyPEM)
	if err != nil {
		return nil, apperror.Unauthorized("jwt_invalid")
	}

	token, err := jwt.Parse(tokenStr, func(t *jwt.Token) (interface{}, error) {
		if _, ok := t.Method.(*jwt.SigningMethodRSA); !ok {
			return nil, fmt.Errorf("unexpected signing method")
		}
		return pubKey, nil
	}, jwt.WithAudience(identifier))

	if err != nil || !token.Valid {
		return nil, apperror.Unauthorized("jwt_invalid")
	}

	claims, ok := token.Claims.(jwt.MapClaims)
	if !ok {
		return nil, apperror.Unauthorized("jwt_invalid")
	}

	iss, _ := claims["iss"].(string)
	if iss != uc.cfg.JWT.Issuer {
		return nil, apperror.Unauthorized("jwt_invalid")
	}

	return map[string]interface{}(claims), nil
}

// parseRSAPrivateKey はPEM文字列からRSA秘密鍵を解析します。
func parseRSAPrivateKey(pemStr string) (*rsa.PrivateKey, error) {
	block, _ := pem.Decode([]byte(pemStr))
	if block == nil {
		return nil, fmt.Errorf("failed to decode PEM")
	}
	return x509.ParsePKCS1PrivateKey(block.Bytes)
}

// parseRSAPublicKey はPEM文字列からRSA公開鍵を解析します。
func parseRSAPublicKey(pemStr string) (*rsa.PublicKey, error) {
	block, _ := pem.Decode([]byte(pemStr))
	if block == nil {
		return nil, fmt.Errorf("failed to decode PEM")
	}
	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, err
	}
	rsaPub, ok := pub.(*rsa.PublicKey)
	if !ok {
		return nil, fmt.Errorf("not RSA public key")
	}
	return rsaPub, nil
}
